Add ScalarTypeStrategy

R4c00n opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description

We quite often have the case that we need to do a type cast before we can hydrate a value.

The `TypeCastStrategy` provides named constructors to cast a value to `int`, `float` or `string`.

Do you think such a feature could be useful directly in this project? And if so, would you prefer a single strategy (like the current state of this PR) or multiple ones, like `IntStrategy`, `FloatStrategy` and `StringStrategy`?

Commit:
Add TypeCastStrategy

Signed-off-by: Marcel Kempf <marcel.kempf@check24.de>
Signed-off-by: Marcel Kempf <marcel.kempf93@gmail.com>

samsonasik commented on src/Strategy/TypeCastStrategy.php:
```diff
@@ -0,0 +1,76 @@
+     */
+    public function hydrate($value, ?array $data)
+    {
+        return call_user_func_array($this->type . 'val', [$value]);
```
can use direct call, ref https://3v4l.org/YdY9t 

```suggestion
        return ($this->type . 'val')($value);
```

Actually, it seems `(int)`, `(float)`, and `(string)` can be used instead ?

R4c00n replied:
Nice, didn't know this, thanks!
Yes, this could also be done with a switch/case on `$this-type`. I'm fine with either way. What do you prefer?

samsonasik replied:
switch/case seems ok for me.

R4c00n committed:
Invoke instead of call_user_func_array

Co-authored-by: Abdul Malik Ikhsan <samsonasik@gmail.com>
Signed-off-by: Marcel Kempf <marcel.kempf@check24.de>
Signed-off-by: Marcel Kempf <marcel.kempf93@gmail.com>

Commit:
Use switch-case instead of dynamic cast

Signed-off-by: Marcel Kempf <marcel.kempf@check24.de>
Signed-off-by: Marcel Kempf <marcel.kempf93@gmail.com>

Commit:
Handle null value

Signed-off-by: Marcel Kempf <marcel.kempf@check24.de>
Signed-off-by: Marcel Kempf <marcel.kempf93@gmail.com>

froschdesign commented:
@R4c00n 
> The `TypeCastStrategy` provides named constructors to cast a value to `int`, `float` or `string`.

`TypeCast` is too general here because the strategy only supports some _– not all –_ scalar variables. If we combine several types in one strategy, then it should be all of them. This means a new name is required for the strategy like `ScalarTypeStrategy`.
(Correct, a strategy already exists which [converts values into Booleans](https://docs.laminas.dev/laminas-hydrator/v4/strategy/#laminas-hydrator-strategy-booleanstrategy) but both strategies could coexist and with a new major version the duplication can be removed.)

Another option is to detect the types automatically what can be done, for example, with [PHP's validate filters](https://www.php.net/manual/filter.filters.validate.php). _But I don't know if I like it._

I would prefer a hydrator with support for constructor injection that uses the parameter type declarations. :smile:

froschdesign commented on .idea/workspace.xml:
```diff
@@ -0,0 +1,138 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Please remove this file from the pull request. Thanks!

froschdesign commented on src/Strategy/TypeCastStrategy.php:
```diff
@@ -0,0 +1,97 @@
+    /** @var string */
+    private $type;
+
+    /**
+     * @return self<int>
+     */
+    public static function createToInt(): self
+    {
+        return new self(self::TYPE_INT);
+    }
+
+    /**
+     * @return self<float>
+     */
+    public static function createToFloat(): self
+    {
+        return new self(self::TYPE_FLOAT);
+    }
+
+    /**
+     * @return self<string>
+     */
+    public static function createToString(): self
+    {
+        return new self(self::TYPE_STRING);
+    }
```
These methods are not relevant for this hydrator strategy. Please remove them.

R4c00n replied:
Mh, but this would result in something like `new TypeCastStrategy(TypeCaseStrategy::TYPE_STRING)` to create a strategy for a specific type. Or is there another way to create/configure strategies that I'm not aware of?

froschdesign replied:
Ah, I missed that the visibility of the constructor is marked as private.

Commit:
Remove obsolete file

Signed-off-by: Marcel Kempf <marcel.kempf@check24.de>
Signed-off-by: Marcel Kempf <marcel.kempf93@gmail.com>

Commit:
Rename strategy

Signed-off-by: Marcel Kempf <marcel.kempf@check24.de>
Signed-off-by: Marcel Kempf <marcel.kempf93@gmail.com>

R4c00n commented:
@froschdesign Thanks for your input! I changed the name of the strategy to `ScalarTypeStrategy` and also added a boolean type cast, as you suggested.

> (Correct, a strategy already exists which converts values into Booleans but both strategies could coexist and with a new major version the duplication can be removed.)

Should I mark the `BooleanStrategy` as deprecated?

weierophinney committed:
qa: apply laminas-coding-standard v2 rules

Original patch was created before we adopted the newer ruleset.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: fix Psalm failures

- Psalm decided that undeclared static properties declared in the parent class of a PHPUnit test case are the child class's problem.
  Suppress these errors.
- The Psalm templating provided in the original submission was incorrect, as we are using string values to denote the type that the class will respond to.
  This means that Psalm sees a string value always, instead of the intended type.
  I have removed the template and references to it for now.
- The patch was using "str" and "boolean" for the "string" and "bool" types, respectively.
  I've updated the code to reflect accepted naming conventions.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
docs: document new ScalarTypeStrategy, and deprecate BooleanStrategy

- Marks `BooleanStrategy` deprecated both in class annotation and documentation.
- Documents `ScalarTypeStrategy`.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: ignore DeprecatedClass errors in BooleanStrategyTest

- Class is known to be deprecated.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
docs: fix linting errors

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
